compositor_pipeline/src/pipeline/output/whip/establish_peer_connection.rs

```diff
@@ -0,0 +1,205 @@
+
+use super::WhipError;
+use reqwest::{header::HeaderMap, Client, Method, StatusCode, Url};
+use std::sync::{atomic::AtomicBool, Arc};
+use tracing::{debug, error, info, warn};
+use webrtc::{
+    ice_transport::{ice_candidate::RTCIceCandidate, ice_connection_state::RTCIceConnectionState},
+    peer_connection::{sdp::session_description::RTCSessionDescription, RTCPeerConnection},
+    rtcp::payload_feedbacks::picture_loss_indication::PictureLossIndication,
+};
+
+pub async fn connect(
+    peer_connection: Arc<RTCPeerConnection>,
+    endpoint_url: String,
+    bearer_token: Option<String>,
+    should_close: Arc<AtomicBool>,
+    tokio_rt: Arc<tokio::runtime::Runtime>,
+    client: reqwest::Client,
+    request_keyframe_sender: Option<Sender<()>>,
+) -> Result<Url, WhipError> {
+    peer_connection.on_ice_connection_state_change(Box::new(
+        move |connection_state: RTCIceConnectionState| {
+            debug!("Connection State has changed {connection_state}.");
+            if connection_state == RTCIceConnectionState::Connected {
+                debug!("Ice connected.");
+            } else if connection_state == RTCIceConnectionState::Failed {
+                debug!("Ice connection failed.");
+                should_close.store(true, std::sync::atomic::Ordering::Relaxed);
+            }
+            Box::pin(async {})
+        },
+    ));
+
+    if let Some(keyframe_sender) = request_keyframe_sender {
+        let senders = peer_connection.get_senders().await;
+        for sender in senders {
+            let keyframe_sender_clone = keyframe_sender.clone();
+            tokio_rt.spawn(async move {
+                loop {
+                    if let Ok((packets, _)) = &sender.read_rtcp().await {
+                        for packet in packets {
+                            if packet
+                                .as_any()
+                                .downcast_ref::<PictureLossIndication>()
+                                .is_some()
+                            {
+                                if let Err(err) = keyframe_sender_clone.send(()) {
+                                    debug!(%err, "Failed to send keyframe request to the encoder.");
+                                };
+                            }
+                        }
+                    } else {
+                        debug!("Failed to read RTCP packets from the sender.");
+                    }
+                }
+            });
+        }
+    }
+
+    let offer = peer_connection
+        .create_offer(None)
+        .await
+        .map_err(WhipError::OfferCreationError)?;
+
+    let endpoint_url = Url::parse(&endpoint_url)
+        .map_err(|e| WhipError::InvalidEndpointUrl(e, endpoint_url.clone()))?;
+
+    info!("Endpoint url: {}", endpoint_url);
+
+    let mut header_map = HeaderMap::new();
+    header_map.append("Content-Type", "application/sdp".parse().unwrap());
+
+    let bearer_token = bearer_token.map(Arc::new);
```

Comment:
?
Either keep the token as Arc<str> everywhere or as String. It does not make sense to create Arc just to use it latter in `format!`